Remove corrupted WAL files in kPointRecoveryMode with avoid_flush_duing_recovery set true

akankshamahajan15 opened the pull request:
Summary:
 1) In case of non-TransactionDB and avoid_flush_during_recovery = true, RocksDB won't
flush the data from WAL to L0 for all column families if possible. As a
result, not all column families can increase their log_numbers, and
min_log_number_to_keep won't change. 
2) For transaction DB (.allow_2pc), even with the flush, there may be old WAL files that it must not delete because they can contain data of uncommitted transactions and min_log_number_to_keep won't change.

If we persist a new MANIFEST with
advanced log_numbers for some column families, then during a second
crash after persisting the MANIFEST, RocksDB will see some column
families' log_numbers larger than the corrupted wal, and the "column family inconsistency" error will be hit, causing recovery to fail.

As a solution,
1. the corrupted WALs whose numbers are larger than the
corrupted wal and smaller than the new WAL will be moved to archive folder.
2. Currently, RocksDB DB::Open() may creates and writes to two new MANIFEST files even before recovery succeeds. This PR buffers the edits in a structure and writes to a new MANIFEST after recovery is successful

Test Plan: 1. Added new unit tests
                2. make crast_test -j

akankshamahajan15 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1251,13 +1251,40 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
       // log number
       versions_->MarkFileNumberUsed(max_wal_number + 1);
 
+      if (*corrupted_wal_found == true &&
```
TODO: Check if column family log number needs to be advanced or not.

riversand963 commented:
Thanks @akankshamahajan15 for the PR. Hopefully we can improve the recovery process after this series of PRs.

One thing that I should have emphasized: currently, RocksDB will write and persist new MANIFEST files even before recovery succeeds. See https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_open.cc#L611, https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_open.cc#L1280, https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_open.cc#L523, https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_open.cc#L521, etc. (I didn't do a thorough search).

This is behavior is problematic. I think one important thing we want to achieve in this PR is to write new MANIFEST (call LogAndApply) when we are sure the db is consistent.

akankshamahajan15 commented:
> One thing that I should have emphasized: currently, RocksDB will write and persist new MANIFEST files even before recovery succeeds. See https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_open.cc#L611, https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_open.cc#L1280, https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_open.cc#L523, https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl_open.cc#L521, etc. (I didn't do a thorough search).

Yes, you mentioned in the task, but that step wasn't clear to me at that time. I will update the code now to persist new MANIFEST only after the DB is consistent.

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1697,6 +1755,9 @@ Status DBImpl::Open(const DBOptions& db_options, const std::string& dbname,
       }
     }
+    if (s.ok()) {
+      s = impl->LogAndApply(&recovery_version_edits);
```
iiuc, this should happen after the subsequent WAL sync? Is this the issue you mentioned before?

akankshamahajan15 replied:
Yes, It seems like `LogAndApply` needs to be called before `InstallSuperVersionAndScheduleWork` otherwise it fails to find the keys. It didn't get the chance to look into `InstallSuperVersionAndScheduleWork` more but the failing was because of the order of API calls.

riversand963 replied:
Can you try switching the order of writing and syncing WAL with installing superversions as https://github.com/riversand963/rocksdb/blob/tmp-9634/db/db_impl/db_impl_open.cc#L1798:L1810?
Then you can do
- Write and sync WAL
- LogAndApply
- InstallSuperVersions

akankshamahajan15 replied:
Sure. Let me try this one.

akankshamahajan15 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1314,6 +1328,41 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+  size_t i = 0;
+
+  // Find the index of first corrupted wal.
+  while (i < wal_numbers.size() - 1) {
```
Change the condition to (i <= wal_numbers.size()-1)

riversand963 commented on db/corruption_test.cc:
```diff
@@ -912,6 +913,258 @@ TEST_F(CorruptionTest, VerifyWholeTableChecksum) {
+  ColumnFamilyHandle* cfh = nullptr;
+  s = db_->CreateColumnFamily(options, test_cf_name, &cfh);
+  ASSERT_OK(s);
+  if (track_and_verify_wals_in_manifest_) {
+    ASSERT_OK(db_->SyncWAL());
+  }
```
I don't think this is needed.

riversand963 commented on db/corruption_test.cc:
```diff
@@ -912,6 +913,258 @@ TEST_F(CorruptionTest, VerifyWholeTableChecksum) {
+  assert(dbimpl);
+
+  ASSERT_OK(db_->Put(WriteOptions(), handles[1], "dontcare", "dontcare"));
+  if (track_and_verify_wals_in_manifest_) {
+    ASSERT_OK(db_->SyncWAL());
+  }
```
Ditto

riversand963 commented on db/corruption_test.cc:
```diff
@@ -912,6 +913,258 @@ TEST_F(CorruptionTest, VerifyWholeTableChecksum) {
+  options.avoid_flush_during_recovery = avoid_flush_during_recovery_;
+  s = DB::Open(options, dbname_, cf_descs, &handles, &db_);
+  ASSERT_OK(s);
+  if (track_and_verify_wals_in_manifest_) {
+    ASSERT_OK(db_->SyncWAL());
+  }
```
Ditto

riversand963 commented on db/corruption_test.cc:
```diff
@@ -912,6 +913,258 @@ TEST_F(CorruptionTest, VerifyWholeTableChecksum) {
+  handles.clear();
+  CloseDb();
+
+  std::string wal_path = LogFileName(dbname_, max_wal_num);
+  uint64_t old_size = 0;
+  ASSERT_OK(env_->GetFileSize(wal_path, &old_size));
+  assert(old_size > 8);
+  uint64_t new_size = old_size - 8;
+  ASSERT_OK(test::TruncateFile(env_, wal_path, new_size));
```
I think it's helpful to add some comments for this code block, or even put it in a helper function.
The comments should explain what errors we are emulating: power reset which caused us to lose the last un-synced WAL.

riversand963 replied:
Same for the other unit test of transaction db.

riversand963 commented on db/corruption_test.cc:
```diff
@@ -912,6 +913,258 @@ TEST_F(CorruptionTest, VerifyWholeTableChecksum) {
+  handles.clear();
+  CloseDb();
+
+  max_wal_num = 0;
+  files.clear();
+  ASSERT_OK(env_->GetChildren(dbname_, &files));
+  for (const auto& f : files) {
+    uint64_t file_num = 0;
+    FileType type{kWalFile};
+    if (ParseFileName(f, &file_num, &type) && type == kWalFile) {
+      if (file_num > max_wal_num) {
+        max_wal_num = file_num;
+      }
+    }
+  }
+  wal_path = LogFileName(dbname_, max_wal_num);
+  ASSERT_OK(test::TruncateFile(env_, wal_path, new_size = 0));
```
Regarding this block together with the above `DB::Open()`, I think it's helpful to add comments and/or put it in a helper function. The comment should explain what failure case we are emulating: before this PR, we may prematurely persist a new MANIFEST even before we can declare the DB is in consistent state after recovery (this is when the new WAL is synced). Should there be a power failure before we sync the new WAL, we will end up in a situation in which the MANIFEST indicates one or more column family's log number is ahead of the WAL with corruption. Also helpful to explain briefly why the corrupted WAL will still be examined in the next recovery attempt.

riversand963 replied:
Same for the other unit test of transaction db.

akankshamahajan15 replied:
Sure. I will update the comments.

riversand963 commented on db/corruption_test.cc:
```diff
@@ -912,6 +913,258 @@ TEST_F(CorruptionTest, VerifyWholeTableChecksum) {
+                                          std::make_tuple(true, true),
+                                          std::make_tuple(false, true)));
+
+TEST_P(CorruptionTestWithParams1, CrashDuringRecovery) {
```
Nit: we need a better name than `CorruptionTestWIthParams1`.

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1240,6 +1240,38 @@ class DBImpl : public DB {
 
   std::atomic<bool> shutting_down_;
 
+  class RecoveryVersionEdits {
```
This is not just version edits. We can come up with another name, VersionEditsContext, etc?

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1389,6 +1422,8 @@ class DBImpl : public DB {
   Status FailIfTsSizesMismatch(const ColumnFamilyHandle* column_family,
                                const Slice& ts) const;
 
+  Status LogAndApply(RecoveryVersionEdits* recovery_version_edits);
```
Can we name it `LogAndApplyForRecovery`?

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1240,6 +1240,38 @@ class DBImpl : public DB {
+        edit_lists.emplace_back(autovector<VersionEdit*>());
+      }
+      uint32_t i = map[cfd->GetID()];
+      edit_lists[i].emplace_back(new VersionEdit(_edit));
```
When are the objects free'ed?

akankshamahajan15 replied:
In the destructor of this class https://github.com/facebook/rocksdb/pull/9634/files#diff-ce293088d5cc10fe3c3342c0968462651a9a955f6e7d745766b96132b1733801R1245

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1240,6 +1240,38 @@ class DBImpl : public DB {
+      edit_lists[i].emplace_back(new VersionEdit(_edit));
+    }
+
+    std::unordered_map<uint32_t, uint32_t> map;  // cf_id to index;
+    autovector<ColumnFamilyData*> cfds;
+    autovector<const MutableCFOptions*> mutable_cf_opts;
+    autovector<autovector<VersionEdit*>> edit_lists;
```
We should follow https://google.github.io/styleguide/cppguide.html#Variable_Names when naming the variables.

riversand963 commented:
Thinking more about this, I think we do not have to remove the corrupted WALs as long as we persist the new MANIFEST after successfully syncing the new WAL.
- If a future recovery starts from the new MANIFEST, then it means the new WAL is successfully synced. Due to the sentinel empty write batch at the beginning, kPointInTimeRecovery of WAL is guaranteed to go after this point.
- If future recovery starts from the old MANIFEST, it means the writing the new MANIFEST failed. We won't have the "SST ahead of WAL" error.

akankshamahajan15 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1314,6 +1327,51 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+    std::string fname =
+        LogFileName(immutable_db_options_.GetWalDir(), wal_numbers[i]);
+    if (create_status.ok()) {
+      wal_manager_.ArchiveWALFile(fname, wal_numbers[i]);
```
@riversand963 I moved the files to an archive folder and if for some reason it fails to create folder, it deletes the files. Do you think it's too aggressive or it's fine? If for some reason folder is not created, it will generate inconsistency error when open after recovery.

riversand963 replied:
let's not delete the WALs in this case. We should only delete the WAL once DB figures out the WAL is not needed (< min_log_number_to_keep).

akankshamahajan15 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1314,6 +1327,51 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+  // (max_wal_number) to avoid column family inconsistency error to archival
+  // directory. If its unable to create archive dir, it will delete the
+  // corrupted WAL files.
+  while (i <= wal_numbers.size() - 1) {
```
I tried to keep the first corrupted file  but I still get the inconsistency error. So currently I moved all corrupted files to an archive folder.

riversand963 replied:
Weird. `kPointInTimeRecovery` should be able to proceed to the most-recently-created-and-synced WAL.
What is the cause for the failure?

riversand963 replied:
Any findings?

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1256,53 +1269,53 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+              WALRecoveryMode::kPointInTimeRecovery) {
+        MoveCorruptedWalFiles(wal_numbers, corrupted_wal_number);
+        // Since last log has been deleted, no need to truncate the log.
+        truncate_last_log = false;
```
Can we just truncate the last WAL? From the implementation of `GetLogSizeAndMaybeTruncate()`, it truncates the file to its actual size to reclaim pre-allocated spaces.

riversand963 replied:
Since there have been quite a number of comments, just want to make sure that we still consider this one. :)

akankshamahajan15 replied:
Yes, we do.  Before moving  the corrupted WAL, it truncates the last WAL to reclaim the pre allocated space before.

riversand963 reviewed (commented):
Thanks @akankshamahajan15 for tackling this tricky bug, and we are getting close.
Left some minor and cosmetic comments.

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1314,6 +1327,62 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+  size_t i = 0;
+  size_t num_wals = wal_numbers.size();
+  // Find the index of first corrupted wal.
+  while (i < num_wals) {
```
Looks like we can assert the `wal_numbers` is sorted.

riversand963 replied:
if sorted, use binary search

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1314,6 +1327,62 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+  // directory. If its unable to create archive dir, it will delete the
+  // corrupted WAL files.
+  // We are moving all but first corrupted WAL file to a different folder.
+  while (i < num_wals) {
```
iiuc, the unit tests we added so far does not cover the case of (i < num_wals) because the last WAL is corrupted? so there is really no WAL between the first corrupted WAL and the new WAL created during recovery.

akankshamahajan15 commented on db/db_impl/db_impl_files.cc:
```diff
@@ -892,20 +892,28 @@ Status DBImpl::SetDBId(bool read_only) {
-                                 mutable_cf_options, &edit, &mutex_, nullptr,
-                                 /* new_descriptor_log */ false);
+      if (version_edits_ctx == nullptr) {
+        // In case of ReadOnly DB, recovery_version_edits is nullptr.
```
https://github.com/facebook/rocksdb/blob/627deb7ceb732d62df81e283ed425119fadca0ff/db/db_basic_test.cc#L185. This test specifically tests `write_dbid_to_manifest` with `read_only` because of which I added these if else checks. Am I missing something?

riversand963 replied:
Good finding. I think this test was added by mistake. We intentionally pay attention not to create or write to the files if DB is opened for read-only access. One example is https://github.com/facebook/rocksdb/blob/7.1.fb/db/db_impl/db_impl_open.cc#L57. If db is for read-only, we don't even write LOG files. I also remember we have use cases in which application has only read access to the file system.
cc @siying who has more context on the `write_dbid_to_manifest` feature, and @ajkr and @pdillinger who, iirc, have been involved in the discussion of users of `DBImplReadOnly` having only read access to the file system.

pdillinger replied:
I'm not really following the details here, but btw there is a ReadOnlyFileSystem that can be used in testing read-only DB to check that we don't write to FS.

akankshamahajan15 replied:
@pdillinger  The above unit test (in comments above) opens`readonly` db and set the `write_dbid_to_manifest` option to true. So I wanted to confirm do we write db_id to manifest in case of readonly? If not we should remove the unit test.

riversand963 replied:
This is the issue #7966 with more discussions on how we should handle `DBImplReadOnly`.

pdillinger replied:
`write_dbid_to_manifest = true` is not required for the test to pass anyway, so I consider it a low quality test. And I definitely agree that setting `write_dbid_to_manifest = true` should not supersede "read only" open. Ideally we would test that the combination does not actually write to FS, so perhaps the existing (low quality) test can be updated for that (e.g. open read-only with CompositeEnvWrapper+ReadOnlyFileSystem).

akankshamahajan15 replied:
As @riversand963 mentioned,  should we return an error in `ValidateOptions()` if `read_only` and `write_dbid_to_manifest` are both true?

pdillinger replied:
I don't think we need to. (a) It was envisioned that write_dbid_to_manifest=true would be eventual default, so such a limitation would need to be changed if we did that. (b) I think of "read only" as amending the Options, so that you can use the same Options from read-write to open read-only.

akankshamahajan15 commented:
[In Progress] Update the CrashDuringRecoveryWithCorruptionTest unit tests  and ReadOnlyDBWithWriteDBIdToManifestSet

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1360,12 +1392,13 @@ class DBImpl : public DB {
       bool error_if_data_exists_in_wals = false,
-      uint64_t* recovered_seq = nullptr);
+      uint64_t* recovered_seq = nullptr,
+      VersionEditsContext* recovery_version_edits = nullptr);
```
Nit: update the comment to explain `recovery_version_edits`.

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1360,12 +1392,13 @@ class DBImpl : public DB {
 
   // Set DB identity file, and write DB ID to manifest if necessary.
-  Status SetDBId(bool read_only);
+  Status SetDBId(bool read_only, VersionEditsContext* recovery_edit);
```
Nit: update the comment to explain `recovery_edits`.

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1379,7 +1412,7 @@ class DBImpl : public DB {
   // meantime, we find out the largest file number present in the paths, and
   // bump up the version set's next_file_number_ to be 1 + largest_file_number.
-  Status DeleteUnreferencedSstFiles();
+  Status DeleteUnreferencedSstFiles(VersionEditsContext* recovery_edit);
```
Ditto

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1389,6 +1422,8 @@ class DBImpl : public DB {
   Status FailIfTsSizesMismatch(const ColumnFamilyHandle* column_family,
                                const Slice& ts) const;
 
+  Status LogAndApplyForRecovery(VersionEditsContext* recovery_version_edits);
```
Nit: `const(?) VersionEditsContext&`.

riversand963 commented on db/db_impl/db_impl.h:
```diff
@@ -1643,9 +1678,10 @@ class DBImpl : public DB {
                          SequenceNumber* next_sequence, bool read_only,
-                         bool* corrupted_log_found);
+                         bool* corrupted_log_found,
+                         VersionEditsContext* recovery_version_edits);
```
Nit: maybe add an explanation for `recovery_version_edits`?

akankshamahajan15 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -805,10 +808,18 @@ Status DBImpl::InitPersistStatsColumnFamily() {
+  return versions_->LogAndApply(
+      version_edits_ctx->cfds_, version_edits_ctx->mutable_cf_opts_,
+      version_edits_ctx->edit_lists_, &mutex_, directories_.GetDbDir(),
+      /*new_descriptor_log=*/true);
```
Many of the tests are failing if I add `assert(versions_->descriptor_log_ == nullptr)`

riversand963 replied:
Can you share the stack trace? I think one of the purposes of this PR is to make sure this is the first place for RocksDB to write to the MANIFEST after recovery.

akankshamahajan15 replied:
There were many  tests failure. One of them is 
```
FormatDef/ColumnFamilyTest.CreateMissingColumnFamilies/0
column_family_test: db/db_impl/db_impl_open.cc:814: rocksdb::Status rocksdb::DBImpl::LogAndApplyForRecovery(const rocksdb::DBImpl::VersionEditsContext*): Assertion `versions_->descriptor_log_ == nullptr' failed.
Received signal 6 (Aborted)
#0   /usr/local/fbcode/platform009/lib/libc.so.6(gsignal+0x100) [0x7f3507cfe530] ??     ??:0
#1   /usr/local/fbcode/platform009/lib/libc.so.6(abort+0x11c) [0x7f3507ce551c] ??       ??:0
#2   /usr/local/fbcode/platform009/lib/libc.so.6(+0x354c2) [0x7f3507cf54c2] ??  ??:0
#3   /usr/local/fbcode/platform009/lib/libc.so.6(__assert_fail+0x42) [0x7f3507cf5532] ??        ??:0
#4   ./column_family_test() [0x5be276] rocksdb::DBImpl::LogAndApplyForRecovery(rocksdb::DBImpl::VersionEditsContext const*)     /home/akankshamahajan/rocksd/db/db_impl/db_impl_open.cc:814 (discriminator 1)
#5   ./column_family_test() [0x5c4c2b] rocksdb::DBImpl::Open(rocksdb::DBOptions const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<rocksdb::ColumnFamilyDescriptor, std::allocator<rocksdb::ColumnFamilyDescriptor> > const&, std::vector<ro ksdb::ColumnFamilyHandle*, std::allocator<rocksdb::ColumnFamilyHandle*> >*, rocksdb::DB**, bool, bool)        /home/akankshamahajan/rocksd/db/db_impl/db_impl_open.cc:1899
#6   ./column_family_test() [0x5c5805] rocksdb::DB::Open(rocksdb::DBOptions const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<rocksdb::ColumnFamilyDescriptor, std::allocator<rocksdb::ColumnFamilyDescriptor> > const&, std::vector<rocksd ::ColumnFamilyHandle*, std::allocator<rocksdb::ColumnFamilyHandle*> >*, rocksdb::DB**)        /home/akankshamahajan/rocksd/db/db_impl/db_impl_open.cc:1633
#7   ./column_family_test() [0x4c5575] rocksdb::ColumnFamilyTestBase::TryOpen(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<rocksdb::C lumnFamilyOptions, std::allocator<rocksdb::ColumnFamilyOptions> >)    /home/akankshamahajan/rocksd/db/column_family_test.cc:186
#8   ./column_family_test() [0x49de0a] rocksdb::ColumnFamilyTest_CreateMissingColumnFamilies_Test::TestBody()   /home/akankshamahajan/rocksd/db/column_family_test.cc:2235 (discriminator 4)
#9   ./column_family_test() [0x5265b6] void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)    /home/akankshamahajan/rocksd/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3899
#10  ./column_family_test() [0x51d6b6] testing::Test::Run()     /home/akankshamahajan/rocksd/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3973
#11  ./column_family_test() [0x51d8f4] testing::Test::Run()     /home/akankshamahajan/rocksd/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3965
#12  ./column_family_test() [0x51da84] testing::TestInfo::Run() /home/akankshamahajan/rocksd/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:4124
#13  ./column_family_test() [0x51e0f1] testing::TestCase::Run() /home/akankshamahajan/rocksd/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:4253
#14  ./column_family_test() [0x526a86] bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)      /home/akankshamahajan/rocksd/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3899
#15  ./column_family_test() [0x51e245] testing::UnitTest::Run() /home/akankshamahajan/rocksd/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:6242
#16  ./column_family_test() [0x44eba8] RUN_ALL_TESTS()  /home/akankshamahajan/rocksd/third-party/gtest-1.8.1/fused-src/gtest/gtest.h:22110
#17  /usr/local/fbcode/platform009/lib/libc.so.6(__libc_start_main+0x104) [0x7f3507ce5dc4] ??   ??:0
#18  ./column_family_test() [0x46eb6d] _start   ??:?

```

akankshamahajan15 replied:
Detailed tests: https://app.circleci.com/pipelines/github/facebook/rocksdb/12584/workflows/b313066b-1dfc-4b3e-8b1b-f185e30a94df/jobs/334412

akankshamahajan15 replied:
I will check the test and debug to see where descriptor_log_ is not nullptr.

riversand963 replied:
That will be great! Thanks

akankshamahajan15 replied:
@riversand963 I looked into one of the test `CreateMissingColumnFamily` https://github.com/facebook/rocksdb/blob/bbcf7b192c12f4ad99e57f5727dd8e1429e206b3/db/column_family_test.cc#L2231-L2237. 

The order is:
 `DB::Open` -> 
`DBImpl::CreateColumnFamily` ->(https://github.com/facebook/rocksdb/blob/bbcf7b192c12f4ad99e57f5727dd8e1429e206b3/db/db_impl/db_impl_open.cc#L1772)  
`VersionSet::LogAndApply ` (https://github.com/facebook/rocksdb/blob/bbcf7b192c12f4ad99e57f5727dd8e1429e206b3/db/db_impl/db_impl.cc#L2768)

akankshamahajan15 replied:
For another test in stats_history_test  `GetStatsHistoryFromDisk ` https://github.com/facebook/rocksdb/blob/bbcf7b192c12f4ad99e57f5727dd8e1429e206b3/monitoring/stats_history_test.cc#L288

`ReopenWithColumnFamilies` -> .... 
`DBImpl::Recover` ->
`DBImpl::InitPersistStatsColumnFamily` -> (https://github.com/facebook/rocksdb/blob/bbcf7b192c12f4ad99e57f5727dd8e1429e206b3/db/db_impl/db_impl_open.cc#L540)
`DBImpl::CreateColumnFamily` ->
`DBImpl::CreateColumnFamilyImpl` ->
`VersionSet::LogAndApply`

akankshamahajan15 replied:
Weirdly `new_descriptor_log` is passed false but still `versions_->descriptor_log_` is not nullptr. Let me further look into it.

akankshamahajan15 replied:
Finally I got it 😅 https://github.com/facebook/rocksdb/blob/8ce7cea93f8c1a8514b1affe84944eb58d545aaa/db/version_set.cc#L4370
This line sets `new_descriptor_log` to true when we first call `LogAndApply`

akankshamahajan15 commented:
> [In Progress] Update the CrashDuringRecoveryWithCorruptionTest unit tests and ReadOnlyDBWithWriteDBIdToManifestSet

Updated `CrashDuringRecoveryWithCorruptionTest`

riversand963 commented:
> in transaction DB, RocksDB won't flush the data from WAL to L0 for all column families if possible

in transaction DB (.allow_2pc), we actually sanitize `avoid_flush_during_recovery` to `false`, thus RocksDB still flushes WAL to L0. Even so, there may be old WAL files that we must not delete because they can contain data of uncommitted transactions.

riversand963 commented on monitoring/stats_history_test.cc:
```diff
@@ -604,10 +604,11 @@ TEST_F(StatsHistoryTest, ForceManualFlushStatsCF) {
       [&] { mock_clock_->MockSleepForSeconds(kPeriodSec); });
   // writing to all three cf, flush default cf
-  // LogNumbers: default: 14, stats: 4, pikachu: 4
+  // LogNumbers: default: 16, stats: 10, pikachu: 5
+  // Since in recovery process, cfd_stats column is created after WAL is
+  // created, synced and MANIFEST is persisted, its log number which depends on
+  // logfile_number_ will be different.
```
Nit: in this test, cf "pikachu" is never flushed, thus its log_number should be the smallest of the three. We can simply assert on this.
```
ASSERT_LT(cfd_test->GetLogNumber(), cfd_stats->GetLogNumber());
ASSERT_LT(cfd_test->GetLogNumber(), cfd_default->GetLogNumber());
```

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1314,6 +1329,50 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+  size_t num_wals = wal_numbers.size();
+
+  // Find the index of first corrupted wal.
+  size_t i = static_cast<size_t>(std::lower_bound(wal_numbers.begin(),
```
Mixing integral vector index with iterator seems odd. Can we just use iterator throughout this function?

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1828,6 +1862,52 @@ Status DBImpl::Open(const DBOptions& db_options, const std::string& dbname,
+    s = impl->LogAndApplyForRecovery(&version_edits_ctx);
+  }
+
+  // DB mutex is already held
```
Nit: replace this assertion with `mutex_.AssertHeld()`.

riversand963 reviewed (commented):
Thanks a lot @akankshamahajan15! Left some additional comments.
After addressing them, we can run stress tests on dev-server and make sure it passes, as part of test plan.

akankshamahajan15 commented on db/db_impl/db_impl.h:
```diff
@@ -1240,6 +1240,40 @@ class DBImpl : public DB {
+  // with corresponding column_family_data and column_family_options.
+  class VersionEditsContext {
+   public:
+    ~VersionEditsContext() {
```
We are not initializing struct members so compiler's default constructor seems ok.

akankshamahajan15 commented:
> Thanks a lot @akankshamahajan15! Left some additional comments. After addressing them, we can run stress tests on dev-server and make sure it passes, as part of test plan.

I ran make crash_test -j for its entire duration and it completed successfully.

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1314,6 +1342,43 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+  // Increment iter to move WAL files from first corrupted_wal_number + 1.
+  iter++;
+
+  std::string archivalPath =
```
Nit: `archival_path`.

riversand963 commented on db/db_impl/db_impl_open.cc:
```diff
@@ -1314,6 +1334,47 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+
+  // Truncate the last WAL to reclaim the pre allocated space before
+  // moving it.
+  GetLogSizeAndMaybeTruncate(wal_numbers.back(), /*truncate=*/true, nullptr);
```
Check return status or call PermitUncheckedError().

riversand963 reviewed (approved):
LGTM. Thanks @akankshamahajan15

riversand963 commented:
Also, can you update HISTORY.md?

akankshamahajan15 commented:
> Also, can you update HISTORY.md?

I updated the HISTORY.md. I added two separate fixes. One is moving corruption wals. Second is to create only one MANIFEST.

ajkr commented:
@akankshamahajan15 Is this right?

> the corrupted WALs whose numbers are larger than the
> corrupted wal and smaller than the new WAL will be moved to archive folder.

In the following recovery we archived (126119, 126216].

```
2022/04/20-20:12:59.858251 556 [db/db_impl/db_impl_open.cc:1163] Point in time recovered to log #126119 seq #1362971066
2022/04/20-20:12:59.858951 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126125 mode 2
2022/04/20-20:13:02.307344 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126128 mode 2
2022/04/20-20:13:05.299616 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126137 mode 2
2022/04/20-20:13:06.112663 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126142 mode 2
2022/04/20-20:13:06.807948 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126145 mode 2
2022/04/20-20:13:06.922859 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126151 mode 2
2022/04/20-20:13:08.407391 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126154 mode 2
2022/04/20-20:13:12.479291 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126160 mode 2
2022/04/20-20:13:13.141196 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126165 mode 2
2022/04/20-20:13:13.689726 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126168 mode 2
2022/04/20-20:13:14.429917 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126170 mode 2
2022/04/20-20:13:14.791170 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126172 mode 2
2022/04/20-20:13:15.354656 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126178 mode 2
2022/04/20-20:13:17.160268 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126183 mode 2
2022/04/20-20:13:19.541990 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126193 mode 2
2022/04/20-20:13:20.461264 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126196 mode 2
2022/04/20-20:13:21.247930 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126202 mode 2
2022/04/20-20:13:23.510047 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126205 mode 2
2022/04/20-20:13:24.541666 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126211 mode 2
2022/04/20-20:13:26.070769 556 [db/db_impl/db_impl_open.cc:913] Recovering log #126216 mode 2
```

Then the following recovery processes 126119 followed by 126219, which doesn't have consecutive seqnos so recovery terminates at 126119:

```
2022/04/20-23:38:04.973200 548 [db/db_impl/db_impl_open.cc:1163] Point in time recovered to log #126119 seq #1362971066
2022/04/20-23:38:04.973920 548 [db/db_impl/db_impl_open.cc:913] Recovering log #126219 mode 2
2022/04/20-23:38:05.117620 548 [WARN] [db/db_impl/db_impl_open.cc:919] 126219.log: dropping 47392297 bytes
2022/04/20-23:38:05.118405 548 [ERROR] [db/db_impl/db_impl_open.cc:1217] Column family inconsistency: SST file contains data beyond the point of corruption.
```

ajkr commented:
BTW, I still haven't figured out why we need to remove the corrupt WALs rather than skip over them during recovery, so any clarification on that would be appreciated.

riversand963 commented:
> BTW, I still haven't figured out why we need to remove the corrupt WALs rather than skip over them during recovery, so any clarification on that would be appreciated.

We can but we do not have to.

> Thinking more about this, I think we do not have to remove the corrupted WALs as long as we persist the new MANIFEST after successfully syncing the new WAL.

> If a future recovery starts from the new MANIFEST, then it means the new WAL is successfully synced. Due to the sentinel empty write batch at the beginning, kPointInTimeRecovery of WAL is guaranteed to go after this point.
If future recovery starts from the old MANIFEST, it means the writing the new MANIFEST failed. We won't have the "SST ahead of WAL" error.

riversand963 commented:
Wait, I remember something, though.
It seems we only sync the new WAL with empty sentinel batch during recovery. It seems possible for an unsynced WAL before the new WAL to become corrupted, without violating the contract between RocksDB and the FS.
Assume we have 10.log, 11.log and 12.log. None of them has been sync'ed, and all of them are higher than "min_log_number_to_keep".
In the first recovery, we stop at 11.log, but in second attempt, is it possible to find a corruption in 10.log which causes the sequence numbers of 10.log and 11.log not consecutive? The WALs are not sync'ed.
